nucypher/blockchain/eth/actors.py

```diff
@@ -347,43 +347,106 @@ def deploy_beneficiary_contracts(self,
-            except TransactionFailed:
-                if crash_on_failure:
-                    raise
-                self.log.debug(f"Failed allocation transaction for {allocation['amount']} to {allocation['beneficiary_address']}")
-                failed.append(allocation)
-                continue
+        if interactive:
+            click.confirm("Continue with the allocation process?", abort=True)
 
-            else:
-                allocation_txhashes[allocation['beneficiary_address']] = txhashes
+        total_to_allocate = NU.from_nunits(sum(allocation['amount'] for allocation in allocations))
+        balance = ContractAgency.get_agent(NucypherTokenAgent, self.registry).get_balance(self.deployer_address)
+        if balance < total_to_allocate:
+            raise ValueError("Not enough tokens to allocate. We need at least {total_to_allocate}.")
+
+        allocation_receipts, failed, allocated = dict(), list(), list()
+        total_deployment_transactions = len(allocations) * 4
+
+        with click.progressbar(length=total_deployment_transactions,
+                               label="Allocation progress",
+                               show_eta=False) as bar:
+            bar.short_limit = 0
+            for allocation in allocations:
+
+                # TODO: Check if allocation already exists in allocation registry
 
-        if failed:
-            # TODO: More with these failures: send to isolated logfile, and reattempt
-            self.log.critical(f"FAILED TOKEN ALLOCATION - {len(failed)} Allocations failed.")
+                beneficiary = allocation['beneficiary_address']
+                name = allocation.get('name', 'No name provided')
 
-        return allocation_txhashes
+                if interactive:
+                    click.pause(info=f"\nPress any key to continue with allocation for "
+                                     f"beneficiary {beneficiary} ({name})")
+
+                if emitter:
+                    emitter.echo(f"\nDeploying UserEscrow contract for beneficiary {beneficiary} ({name})...")
+                    bar._last_line = None
+                    bar.render_progress()
+
+                deployer = self.deploy_user_escrow(allocation_registry=allocation_registry,
+                                                   progress=bar)
+
+                amount = allocation['amount']
+                duration = allocation['duration_seconds']
+                try:
+                    receipts = deployer.deliver(value=amount,
+                                                duration=duration,
+                                                beneficiary_address=beneficiary,
+                                                progress=bar)
+                except TransactionFailed:
+                    if crash_on_failure:
+                        raise
+                    self.log.debug(f"Failed allocation transaction for {NU.from_nunits(amount)} to {beneficiary}")
```

Comment:
Do we want to include the error message from TransactionFailed in the debug message?

Reply:
Good idea.

Reply:
Hmm - Not sure I agree - Its a very generic message, no?

Reply:
As far as I remember, those messages are usually of very little help. In any case, I don't think it can hurt to add them.

Reply:
✅